File: src/test/java/leets/leenk/domain/feed/domain/service/CommentGetServiceTest.java
package leets.leenk.domain.feed.domain.service;

import leets.leenk.domain.feed.application.exception.CommentNotFoundException;
import leets.leenk.domain.feed.domain.entity.Comment;
import leets.leenk.domain.feed.domain.entity.Feed;
import leets.leenk.domain.feed.domain.repository.CommentRepository;
import leets.leenk.domain.feed.test.CommentTestFixture;
import leets.leenk.domain.feed.test.FeedTestFixture;
import leets.leenk.domain.feed.test.UserTestFixture;
import leets.leenk.domain.user.domain.entity.User;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;

import java.util.List;
import java.util.Optional;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;
import static org.mockito.BDDMockito.given;
import static org.mockito.BDDMockito.then;

@ExtendWith(MockitoExtension.class)
public class CommentGetServiceTest {
    @Mock
    private CommentRepository commentRepository;

    @InjectMocks
    private CommentGetService commentGetService;

    @Test
    @DisplayName("findCommentByIdNotDeleted 테스트")
    void findCommentByIdNotDeleted1() {
        //given
        User user = UserTestFixture.createUser(1L, "me");
        Feed feed = FeedTestFixture.createFeed(1L, user);
        Comment comment = CommentTestFixture.createComment(12L, user, feed, "hi");

        given(commentRepository.findByCommentIdAndDeletedAtIsNull(12L))
                .willReturn(Optional.of(comment));

        // when
        Comment result = commentGetService.findCommentByIdNotDeleted(12L);

        // then
        assertThat(result).isSameAs(comment);
        then(commentRepository).should().findByCommentIdAndDeletedAtIsNull(12L);
    }

    @Test
    @DisplayName("findCommentByIdNotDeleted 테스트 - 예외")
    void findCommentByIdNotDeleted2() {
        // given
        given(commentRepository.findByCommentIdAndDeletedAtIsNull(99L))
                .willReturn(Optional.empty());

        // when & then
        assertThatThrownBy(() -> commentGetService.findCommentByIdNotDeleted(99L))
                .isInstanceOf(CommentNotFoundException.class);
        then(commentRepository).should().findByCommentIdAndDeletedAtIsNull(99L);
    }

    @Test
    @DisplayName("findAllByFeed 테스트")
    void findAllByFeed() {
        // given
        User user = UserTestFixture.createUser(1L, "me");
        Feed feed = FeedTestFixture.createFeed(1L, user);
        Comment first = CommentTestFixture.createComment(1L, user, feed, "first");
        Comment second = CommentTestFixture.createComment(2L, user, feed, "second");

        given(commentRepository.findAllByFeedAndDeletedAtIsNullOrderByCreateDateDesc(feed))
                .willReturn(List.of(first, second));

        // when
        List<Comment> result = commentGetService.findAllByFeed(feed);

        // then
        assertThat(result).containsExactly(first, second);
        then(commentRepository).should().findAllByFeedAndDeletedAtIsNullOrderByCreateDateDesc(feed);
    }
}

File: src/test/kotlin/leets/leenk/domain/feed/domain/service/CommentGetServiceTest.kt
package leets.leenk.domain.feed.domain.service

import io.kotest.assertions.throwables.shouldThrow
import io.kotest.core.spec.style.BehaviorSpec
import io.kotest.matchers.collections.shouldContainExactly
import io.kotest.matchers.shouldBe
import io.mockk.every
import io.mockk.mockk
import io.mockk.verify
import leets.leenk.domain.feed.application.exception.CommentNotFoundException
import leets.leenk.domain.feed.domain.repository.CommentRepository
import leets.leenk.domain.feed.test.CommentTestFixture
import leets.leenk.domain.feed.test.FeedTestFixture
import leets.leenk.domain.feed.test.UserTestFixture
import java.util.Optional

class CommentGetServiceTest :
    BehaviorSpec({
        val commentRepository = mockk<CommentRepository>()
        val commentGetService = CommentGetService(commentRepository)

        Given("삭제되지 않은 댓글이 존재할 때") {
            val user = UserTestFixture.createUser(1L, "me")
            val feed = FeedTestFixture.createFeed(1L, user)
            val comment = CommentTestFixture.createComment(12L, user, feed, "hi")

            every { commentRepository.findByCommentIdAndDeletedAtIsNull(12L) } returns Optional.of(comment)

            When("ID로 댓글을 조회하면") {
                val result = commentGetService.findCommentByIdNotDeleted(12L)

                Then("해당 댓글이 반환되어야 한다") {
                    result shouldBe comment
                    verify { commentRepository.findByCommentIdAndDeletedAtIsNull(12L) }
                }
            }
        }

        Given("삭제되지 않은 댓글이 존재하지 않을 때") {
            every { commentRepository.findByCommentIdAndDeletedAtIsNull(99L) } returns Optional.empty()

            When("ID로 댓글을 조회하면") {
                Then("CommentNotFoundException이 발생해야 한다") {
                    shouldThrow<CommentNotFoundException> {
                        commentGetService.findCommentByIdNotDeleted(99L)
                    }
                    verify { commentRepository.findByCommentIdAndDeletedAtIsNull(99L) }
                }
            }
        }

        Given("피드에 여러 댓글이 존재할 때") {
            val user = UserTestFixture.createUser(1L, "me")
            val feed = FeedTestFixture.createFeed(1L, user)
            val first = CommentTestFixture.createComment(1L, user, feed, "first")
            val second = CommentTestFixture.createComment(2L, user, feed, "second")

            every { commentRepository.findAllByFeedAndDeletedAtIsNullOrderByCreateDateDesc(feed) } returns
                listOf(first, second)

            When("피드의 모든 댓글을 조회하면") {
                val result = commentGetService.findAllByFeed(feed)

                Then("생성일 역순으로 정렬된 댓글 목록이 반환되어야 한다") {
                    result shouldContainExactly listOf(first, second)
                    verify { commentRepository.findAllByFeedAndDeletedAtIsNullOrderByCreateDateDesc(feed) }
                }
            }
        }
    })
